keras/engine/training.py

```diff
@@ -2331,6 +2334,15 @@ def evaluate_generator(self, generator, steps=None,
         """
         self._make_test_function()
 
+        stateful_metric_indices = []
```

Comment:
This reset is needed to make Stateful Metrics work for generators.

How do you feel about spinning this bug fix out into a separate PR? Should be a quick approval.

Some of the other changes, for instance ``m.stateful`` will likely have some discussion. I really want this bug fix to make it into the next release :)

Reply:
Do you mind if I raise the reset in evaluate generator to get it through?

Don't want to steal your thunder.

Reply:
No worries about that! But I'm going to take better care of the MR from now on and update it in a couple of minutes, sorry about the delays!

Reply:
Looks like you got it, I just wanted to make sure this PR wasn't going to get stuck for another couple of releases.